python-checks/src/main/java/org/sonar/python/checks/FunctionReturnTypeCheck.java

```diff
@@ -124,6 +128,22 @@ private static void raiseIssues(SubscriptionContext ctx, FunctionDef functionDef
     });
   }
 
+  private static boolean canBeOrExtendIterableType(InferredType inferredType) {
+    return Stream.concat(ITERABLE_TYPES.stream(), ASYNC_ITERABLE_TYPES.stream())
+      .anyMatch(typeName -> canBeOrExtend(inferredType, typeName));
+  }
+
+  private static Boolean canBeOrExtend(InferredType inferredType, String typeName) {
```

Comment:
Do we need to use the boxed type here?

Also I'm a bit uncomfortable with the naming here, or at least with the fact that we still have a `DeclaredType#canBeOrExtend` method with a different implementation.

It would be nice to have a comment explaining that in this case, we consider the declared type of the type annotation to be as trustworthy as if it were a runtime type (which really is the reason for the different implementation).

Maybe a different name would be nice as well, but I know naming is all over the place for these methods and I don't have a good suggestion right now.